pkg/controller/replica.go

```diff
@@ -162,9 +165,31 @@ func (r *ReconcileReplica) triggerGarbageCollection(volumeName string) {
 func (r *ReconcileReplica) removeGarbageCollection(volumeName string) {
 	v, ok := r.cleanUpMap.LoadAndDelete(volumeName)
 	if ok {
-		cancelFunc := v.(context.CancelFunc)
-		cancelFunc()
+		entry := v.(cleanUpEntry)
+		entry.cancelFunc()
+	}
+}
+
+func (r *ReconcileReplica) retryManageReplicas(ctx context.Context, volumeName string) {
+	// retry replica management with exponential backoff until success
+	conditionFunc := func() (bool, error) {
+		// we should not run replica management while garbage collection is scheduled for the volume
+		if v, cleanUpScheduled := r.cleanUpMap.Load(volumeName); cleanUpScheduled {
+			entry := v.(cleanUpEntry)
+			select {
+			case <-entry.ctx.Done():
+				// if gargage collection cancelled, start replica management
+				break
+			case <-entry.initChannel:
```

Comment:
Since there's no `default` case, this `select` statement will block if both `ctx.Done()` and `initChannel` are open with no sends pending. Since another goroutine is likely waiting on `initChannel` in this case and only one goroutine (including the current one) will be selected when the timer expiration is sent, the current goroutine may consume that notification. In that case, the other goroutine will be leaked and neither replica management nor cleanup will be performed. I do not believe that simply adding a `default` case will prevent the race but make it less likely to happen.

Reply:
Good catch! Just submitted another commit to address this 👍